### ISSUE TITLE
The anti-clustering effect in GMVAE

### ISSUE THREAD

**Issue by hammurabi-ds** [psanch21/VAE-GMVAE#1]
Hey psanch21

How did you overcome the issue mentioned by the authors in the paper on the over-regularisation of the y-prior? my model merges all the clusters into one, i cant see how you avoided that in your code. I also have question regarding the modifications you have made to their optimisation, can you share the ideas?

**Comment by psanch21** [comment]
Hi! Regarding your first question we do not overcome the issue for over-regularization of the y-prior. As they stated in the paper,  though there are several approach to mitigate over-regularization, it is still an open problem today.

 In relation to your second question, our implementation differs from the original one in the approximation to the conditional prior term.  If you take a look at equation (5) of the original paper, the latent variable w is sampled from q(w|y) but it is not clear (to me) how they sample x to obtain the final approximation. In our approximation, I think is it more clear since we sample w from  q(w|y)  and we sample x from  q(x|y).

Note: be aware in the description of the formulas in this repository we have renamed the variables of the original paper as follows: z --> y, y --> x, x --> z.

I hope this helps!

**Comment by hammurabi-ds** [comment]
Great thank you for the response. I can see it makes a bit more sense now. I actually did notice their unusual naming of the variables. I still have one final question. They mention a trick to deactivate the y-prior term since the algo had the tendency to initially merge all clusters into one and not getting out of this later on, so instead of the usual y-prior:

![image](https://user-images.githubusercontent.com/32635777/51083484-7837f600-171b-11e9-8819-8d982e0ec4dc.png)

They use: 

![image](https://user-images.githubusercontent.com/32635777/51083491-85ed7b80-171b-11e9-8f47-14b7f07382ef.png)

I can see that you didn't include this term in your loss function, so how did your algo manage to overcome this problem and find the right clusters?

**Comment by psanch21** [comment]
Hi. First of all, I'll like to apologise for the delay in the response. You are right, we did not include that term in the loss function. We actually found that issue with KL of the discrete variable in some experiments. What we did to mitigate this problem was to try different initialisations and use a beta parameter in the y prior term as follows `y_term_new = beta*y_term`  where beta is computed as `beta= min(0.01*epochs, 1)` . In this way, we try to avoid  over-regularisation in the early training stages.

**Comment by hammurabi-ds** [comment]
Hi again, no worries and thank you for the response.  Yes i actually tried out with the this beta term usually called warm up period. Unfortunately this model appears very unstable for my type of data which is very large feature size and low sample size.  

I suggest you have a look at a similar but simpler architecture model with only one latent variable http://ruishu.io/2016/12/25/gmvae/. This have proven to give better performance in the MNIST case. Perhaps for future work :)

**Comment by psanch21** [comment]
Thank you very much for the link, I will take a look at it! Though I have not worked with the type of dataset you described, I think it must be hard to avoid over-fitting using VAE to model them.

**Comment by hammurabi-ds** [comment]
Yes i do have over-fitting problems that i am able to get around to some degree using dropouts, early stopping ects. but its a trade off between being accurate at reconstructing and latent space closeness to unit Gaussian depends on what you want out of the VAE model. Its actually DNA genotype data where you usually have between 500-2000 samples and from 100.000 to 700.000 features.
